Reduce contracts and deployments loading time

samaradel opened the pull request:
### Description

#### Deployments:

- Load all deployments in parallel
- Handle migrations in parallel if needed
- Add grid client cache
- Add batch processing utility
- Create grid clients in parallel with caching
- Load contracts and node IDs in parallel for each machine
- Process consumptions in batches
- Process WireGuard configs in batches

#### Contracts:

- Node info caching
- Batch processing for contract normalization, just like Consumptions and WireGuard
- Parallel data fetching

### Changes

Before: deployments loading time takes around 20 to 25 seconds
After: deployments, loading time takes around 15 to 20 seconds

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/4223

### Tested Scenarios

Check deployments loading `Time taken` in the console for deployments and contracts

Ramez test cases: https://github.com/threefoldtech/tfgrid-sdk-ts/pull/4233#pullrequestreview-3016465988

### Documentation PR

For UI changes, Please provide the Documentation PR on [info_grid](https://github.com/threefoldtech/info_grid)

### To consider

#### Preliminary Checks:

- [x] Preliminary Checks
  - Does it completely address the issue linked?
  - What about edge cases?
  - Does it meet the specified acceptance criteria?
  - Are there any unintended side effects?
  - Does the PR adhere to the team's coding conventions, style guides, and best practices?
  - Does it integrate well with existing features?
  - Does it impact the overall performance of the application?
  - Are there any bottlenecks or slowdowns?
  - Has it been optimized for efficiency?
  - Has it been adequately tested with unit, integration, and end-to-end tests?
  - Are there any known defects or issues?
  - Is the code well-documented?
  - Are changes to documentation reflected in the code?

#### UI Checks:

- [x] UI Checks
  - If a UI design is provided/ does it follow it?
  - Does every button work?
  - Is the data displayed logical? Is it what you expected?
  - Does every validation work?
  - Does this interface feel intuitive?
  - What about slow network? Offline?
  - What if the gridproxy/graphql/chain is failing?
  - What would a first time user have a hard time navigating here?

#### Code Quality Checks:

- [x] Code Quality Checks
  - Code formatted/linted? Are there unused imports? .. etc
  - Is there redundant/repeated code?
  - Are there conditionals that are always true or always false?
  - Can we write this more concisely?
  - Can we reuse this code? If yes, where?
  - Will the changes be easy to maintain and update in the future?
  - Can this code become too complex to understand for other devs?
  - Can this code cause future integration problems?

### Testing Checklist

- [x] Does it Meet the specified acceptance criteria.
- [ ] Test if changes can affect any other functionality.
- [ ] Tested with unit, integration, and end-to-end tests.
- [ ] Tested the un-happy path and rollback scenarios.
- [ ] Documentation updated to meet the latest changes.
- [ ] Check automated tests working successfully with the changes.
- [ ] Can be covered by automated tests.

### General Checklist

- [ ] Tests included
- [x] Build pass
- [ ] Documentation
- [x] Code format and docstring
- [ ] Screenshots/Video attached (needed for UI changes)

Commit:
- Load all deployments in parallel
- Handle migrations in parallel if needed
- Add grid client cache
- Add batch processing utility
- Create grid clients in parallel with caching
- Load contracts and node IDs in parallel for each machine
- Process consumptions in batches
- Process wireguard configs in batches

Commit:
Optimize contracts fetching

Commit:
Merge branch 'development' into dev_load_deployments_performance

Commit:
Remove unused vars

0oM4R commented on packages/playground/src/components/k8s_deployment_table.vue:
```diff
@@ -188,35 +187,82 @@ const loading = ref(false);
 
 onMounted(loadDeployments);
 async function loadDeployments() {
+  const start = performance.now();
```
this is not needed in the code

0oM4R replied:
and pleaes check all other performance related logs

samaradel replied:
I know, it's for you guys to test performance. If it's ok, I'll remove it before merge

0oM4R commented on packages/playground/src/components/k8s_deployment_table.vue:
```diff
@@ -188,35 +187,82 @@ const loading = ref(false);
-  });
-  loading.value = false;
+async function fetchClusterDetails(item: any) {
+  if (item.detailsLoading || (item.billing !== undefined && item.wireguard !== undefined)) return;
```
Why is that needed?

samaradel replied:
To avoid duplicate calls, as it used to fetch billing and wireguard, so it checks if the data already exists ```!== undefined``` will not fetch it again

0oM4R commented on packages/playground/src/components/vm_deployment_table.vue:
```diff
@@ -313,27 +244,23 @@ async function loadDeployments() {
-    const chunk1 = await loadVms(grid!);
-    if (chunk1.count > 0 && migrateGateways) {
-      await migrateModule(grid!.gateway);
+    const [chunk1, chunk2, chunk3] = await Promise.all([
```
I suggest using [allsettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) to avoid errors on partall request error

0oM4R commented on packages/playground/src/utils/load_deployment.ts:
```diff
@@ -35,60 +51,62 @@ async function getDeploymentContracts(grid: GridClient, name: string, projectNam
   });
-  const items = await Promise.all(promises);
+
+  const items = await Promise.all(machinePromises);
```
also here we shouldn't use promise.all

0oM4R commented on packages/playground/src/utils/load_deployment.ts:
```diff
@@ -115,16 +133,27 @@ export async function loadVms(grid: GridClient, options: LoadVMsOptions = {}) {
+
+  const BATCH_SIZE = 10;
+  const consumptions = await batchProcess(vms, BATCH_SIZE, async batch => {
+    return Promise.all(
```
same here should use allsettled

0oM4R reviewed (changes requested):
please use allsettled instead of promise.all for better error handling

Commit:
Use promise.allSettled instead

0oM4R commented on packages/playground/src/weblets/tf_contracts_list.vue:
```diff
@@ -342,16 +340,19 @@ async function loadContractsByType(
-      sortBy: options && options.sort.length ? (options?.sort[0].key as SortByContracts) : undefined,
-      sortOrder: options && options.sort.length ? (options?.sort[0].order as SortOrder) : undefined,
-    });
+    const [response] = await Promise.all([
```
This is still using promise.all please use allsettled

0oM4R commented on packages/playground/src/weblets/tf_contracts_list.vue:
```diff
@@ -317,16 +302,29 @@ async function _normalizeContracts(
+
+  const normalizedContracts: NormalizedContract[] = [];
+  for (const batch of batches) {
+    const results = await Promise.all(
```
also here still using promise.all

0oM4R replied:
@samaradel

0oM4R commented on packages/playground/src/components/k8s_deployment_table.vue:
```diff
@@ -188,35 +187,82 @@ const loading = ref(false);
+      showAllDeployments.value
+        ? loadK8s(updateGrid(grid, { projectName: "" }))
+        : Promise.resolve({ count: 0, items: [], failedDeployments: [] }),
+    ]);
```
please use promise.allsettled

0oM4R commented on packages/playground/src/components/vm_deployment_table.vue:
```diff
@@ -313,27 +244,31 @@ async function loadDeployments() {
+        chunk1.count > 0 && migrateModule(grid!.gateway),
+        chunk2.count > 0 && migrateModule(grid!.gateway),
+        chunk3.count > 0 && migrateModule(grid!.gateway),
+      ]);
```
also here pelsae use promise.allsettled

0oM4R reviewed (commented):
there are many places that are still using promise.all 

please make sure to use promise.allsettled everywhere

0oM4R commented:
there are other places please make sure to check them as well and replace the promise.all with promise.allsettled

Commit:
Replace promise.all accross all deployments files

0oM4R commented on packages/playground/src/weblets/tf_contracts_list.vue:
```diff
@@ -317,16 +302,29 @@ async function _normalizeContracts(
-    }),
-  );
-  return normalizedContracts.filter(Boolean) as NormalizedContract[];
+  const batchSize = 10;
```
For loading batches, the patches will be processed sequentially, and each batch will normalize contracts  in parallel 
first batch-> then second batch

I suggest making the batch size depend on the contract's size to handle even 10 contracts. 
We can divide the contract list size by a constant like 5, so the minimum batch size we have is 5 

Also, we can have outer parallel processing to process the patches in parallel as well.

0oM4R commented on packages/playground/src/weblets/tf_contracts_list.vue:
```diff
@@ -411,6 +426,8 @@ async function loadContracts(type?: ContractType, options?: { sort: { key: strin
     loadingErrorMessage.value = `Error while loading contracts: ${error.message}`;
     createCustomToast(loadingErrorMessage.value, ToastType.danger, {});
   }
+  const end = performance.now();
```
In this function, we can do 
`if (!type) await getTotalCost();`
`await getContractsLockDetails();`
`nodeInfo.value = await getNodeInfo(nodeIDs.value, cachedNodeIDs.value);
    cachedNodeIDs.value.push(...nodeIDs.value);`
    
    In parallel as well as those 3 calls are independent calls

0oM4R commented on packages/playground/src/weblets/tf_contracts_list.vue:
```diff
@@ -522,6 +539,16 @@ async function getContractsLockDetails() {
+    }
+  }
+  return nodeIds.map(id => nodeInfoCache.get(id));
+}
```
- This function should be moved to the utils, imo
- Need to add error handling for getNodeInfo
- I have a concern about invalidating the cache as well

0oM4R commented on packages/playground/src/components/k8s_deployment_table.vue:
```diff
@@ -188,35 +187,88 @@ const loading = ref(false);
+  try {
+    const grid = await getGrid(profileManager.profile!, props.projectName);
+    if (!grid) {
+      loading.value = false;
```
should add error msg here

0oM4R commented on packages/playground/src/components/k8s_deployment_table.vue:
```diff
@@ -188,35 +187,88 @@ const loading = ref(false);
+    if (chunk3.items) {
+      chunk3.items = chunk3.items.map(i => {
+        return !i.projectName || i.projectName === "Kubernetes" ? markAsFromAnotherClient(i) : i;
+      });
```
i don't think returning emtpy result in case of failer is enough, we may log some warning that loading k8s from project name failed

0oM4R commented on packages/playground/src/components/k8s_deployment_table.vue:
```diff
@@ -188,35 +187,88 @@ const loading = ref(false);
+      };
     });
+
+    setTimeout(() => {
+      items.value.forEach(item => fetchClusterDetails(item));
+    }, 0);
```
What is the need for settimeout here?

0oM4R replied:
Also I think those can be loaded in parallel

samaradel replied:
I think it will be good, I'll implement it

0oM4R commented on packages/playground/src/components/k8s_deployment_table.vue:
```diff
@@ -188,35 +187,88 @@ const loading = ref(false);
+      item.detailsLoading = false;
+      return;
+    }
+    const consumption = await grid.contracts.getConsumption({ id: item.masters[0].contractId }).catch(() => undefined);
```
What about passing the grid as arg then using the update grid, 
also here may need to indicate failure

0oM4R commented on packages/playground/src/components/k8s_deployment_table.vue:
```diff
@@ -188,35 +187,88 @@ const loading = ref(false);
+    }
+    const consumption = await grid.contracts.getConsumption({ id: item.masters[0].contractId }).catch(() => undefined);
+    item.billing = consumption ? consumption.amountBilled : "No Data Available";
+    item.wireguard = await grid.networks
```
we can get the consumption and the wireguard in parallel; they are independent calls.

0oM4R commented on packages/playground/src/components/k8s_deployment_table.vue:
```diff
@@ -188,35 +187,88 @@ const loading = ref(false);
+      })
+      .then(res => res[0])
+      .catch(() => undefined);
+  } finally {
```
Please add a catch block; to at least log the errors

samaradel replied:
done

0oM4R commented on packages/playground/src/components/vm_deployment_table.vue:
```diff
@@ -290,7 +220,10 @@ async function loadDomains() {
     const grid = await getGrid(profileManager.profile!, props.projectName.toLowerCase());
     const gateways = await grid!.gateway.list();
-    const gws = await Promise.all(gateways.map(name => grid!.gateway.get_name({ name })));
+    const gwsResults = await Promise.allSettled(gateways.map(name => grid!.gateway.get_name({ name })));
+    const gws = gwsResults
+      .filter(result => result.status === "fulfilled")
+      .map(result => (result as PromiseFulfilledResult<any>).value);
```
need to handle the failure here

0oM4R replied:
@samaradel

samaradel replied:
Already handled in line 229

0oM4R commented on packages/playground/src/components/vm_deployment_table.vue:
```diff
@@ -313,27 +247,31 @@ async function loadDeployments() {
+    const results = await Promise.allSettled([
+      loadVms(grid!),
+      loadVms(updateGrid(grid!, { projectName: props.projectName.toLowerCase() })),
+      showAllDeployments.value && props.projectName.toLowerCase() === ProjectName.VM.toLowerCase()
```
we may group the chunk 3 condition in both files into a separate variable then use it here
example
```ts
loadChunk3 ? loadVms(updateGrid(grid!, { projectName: "" })) :  Promise.resolve({ count: 0, items: [], failedDeployments: [] }),

```

0oM4R commented on packages/playground/src/components/vm_deployment_table.vue:
```diff
@@ -313,27 +247,31 @@ async function loadDeployments() {
+      await Promise.allSettled([
+        chunk1.count > 0 && migrateModule(grid!.gateway),
+        chunk2.count > 0 && migrateModule(grid!.gateway),
+        chunk3.count > 0 && migrateModule(grid!.gateway),
```
i do not understand why we are calling the migrateModule threetimes with the same args, 
also the &&  means if a chunk's count is 0, false is passed to Promise.allSettled() instead of a Promise, which works but isn't ideal.
please use  
```ts
    chunk1.count > 0 ? migrateModule(x) : Promise.resolve(),
```

0oM4R commented on packages/playground/src/utils/batch_process.ts:
```diff
@@ -0,0 +1,13 @@
+    results.push(...batchResults);
+  }
+  return results;
+}
```
we can run the batches in parallel. Also please add error handling and unit test for it

0oM4R reviewed (commented):
partial review will review [packages/playground/src/utils/load_deployment.ts](https://github.com/threefoldtech/tfgrid-sdk-ts/pull/4233/files#diff-b6915db99f68edd26675a9b0555754030c44a2115973d65ef35ef218f5076126)
later

Commit:
- Run the three independent calls  using Promise.allSettled
- Call getTitalCost function only if there is no type defined

Commit:
Fully parallel cluster detail loading

Commit:
- Handle the failure of VM deployments
- Group chunk 3 condition into variable for VMs and K8s
- Improve migrateModule calls

Commit:
- Run batches in parallel
- Handle errors if existed
- Add unit test for batch process function

0oM4R commented on packages/playground/src/utils/batch_process.ts:
```diff
@@ -0,0 +1,63 @@
+      results: batchResults.flat(),
+      errors: [],
+      hasErrors: false,
+    };
```
What is the usage of failFast? and where do we use it please explain

samaradel replied:
well, it was used to control error handling strategy in the process, but it seems it's dead now after the latest changes. I'll remove

0oM4R commented on packages/playground/src/utils/get_nodeInfo_with_cache.ts:
- nodeInfoCache is declared inside the function; it is not persistent between function calls—it's reset (empty) every time the function runs!
also this should be in `get_nodes.ts` not in alone in seperate file

please provide unit test for it as well

0oM4R reviewed (commented):
partal review

Commit:
- Remove dead code
- Remove get_nodeInfo_with_cache file and use it as a seperate function
- Create test unit for getNodeInfoWithCache function
- Extract helper functions to reduce the nesting

Commit:
Fix batch process calling

0oM4R reviewed (commented):
in vm the network got loaded twice, any idea 

<img width="273" height="112" alt="image" src="https://github.com/user-attachments/assets/82b0d361-ffe3-4e24-a962-5e559372f438" />

Commit:
Use cached getGridClient instead of updateGrid to avoid recreating modules

Commit:
Merge branch 'development' into dev_load_deployments_performance

Commit:
Fix conflicts

Commit:
Fix load network duplicates

samaradel commented:
@0oM4R fixed now 
<img width="381" height="236" alt="image" src="https://github.com/user-attachments/assets/fef8f87f-842e-408f-8ede-94bc58092abd" />

ramezsaeed reviewed (commented):
Test Scenarios:
- Test loading contracts time in contract page when have 20 contracts in the account and compare the old time Vs the new time.
- Test deploy a patch VMs and compare time for list deployment and for contract page.
- Deploy a Full VM with all networks enabled and compare listing time Vs new time after this PR. 
- Test if we can have a failed deployment and how the list deployments time will react.

Additional tests:
- Test list deployment time when have VMs in grace period.
- Test list deployment time when have VMs on offline nodes.

0oM4R reviewed (commented):
i have 38 deployment but i can list 41 only 
<img width="1625" height="756" alt="image" src="https://github.com/user-attachments/assets/8b51f0be-eece-45a0-97ff-6cf64ded19d7" />

on https://dashboard.dev.grid.tf/ 
<img width="1596" height="694" alt="image" src="https://github.com/user-attachments/assets/18e10931-cc8a-4747-ab3a-d46d124350cf" />

0oM4R commented:
I tried to make the deployment listing fail 

The error is not accurate, imo
original
<img width="1838" height="782" alt="Pasted Graphic 2" src="https://github.com/user-attachments/assets/e9ef1261-b70e-4f85-bddb-4be031fee918" />

new branch
<img width="1815" height="885" alt="Pasted Graphic 1" src="https://github.com/user-attachments/assets/13e0f1df-8ea1-4f55-a07d-1cc6c2fb3976" />

0oM4R commented:
listing contract have almost no enhancment 
- listing 74 (node only) contract 
 - the original branch 19.877s 
 - the new branch 19.094s
 
 list deployment has a significant enhancement
 
 listing 10 micro vm deployments  
  - original 35.7s
  - old 71.8155s
  jitsi deployments 
  - new branch 43.15s
  - old  75.9s
  
  add 31 name contract so the total now 100 node contract and 31 name contract 
  - new branch 38.211 seconds
  - old one 50.6765s

Commit:
Merge branch 'development' into dev_load_deployments_performance

Commit:
Fix conflicts

ramezsaeed commented:
Tested:
 - Loading time enhanced with 50 VM contracts

samaradel commented:
@0oM4R I see the pagination error and loading deployments timeout exist in development, for the contracts I tested with more than 50 contracts, and it takes 13 seconds
